src/functional/ffront/decorator.py

```diff
@@ -0,0 +1,294 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+import collections
+import dataclasses
+import functools
+import types
+import typing
+from typing import Any, Optional, Protocol
+
+from eve.utils import UIDs
+from functional.common import GTTypeError
+from functional.ffront import common_types as ct
+from functional.ffront import field_operator_ast as foast
+from functional.ffront import program_ast as past
+from functional.ffront import symbol_makers
+from functional.ffront.foast_to_itir import FieldOperatorLowering
+from functional.ffront.func_to_foast import FieldOperatorParser
+from functional.ffront.func_to_past import ProgramParser
+from functional.ffront.past_passes.type_deduction import ProgramTypeDeduction
+from functional.ffront.past_to_itir import ProgramLowering
+from functional.ffront.source_utils import ClosureRefs
+from functional.iterator import ir as itir
+from functional.iterator.backend_executor import execute_program
+
+
+DEFAULT_BACKEND = "roundtrip"
+
+
+@typing.runtime_checkable
+class GTCallable(Protocol):
+    """Protocol (abstract base class) defining the interface for callables.
+
+    Direct subclasses of this base class only need to implement the
+    abstract methods to be callable from inside a program or from other
+    callables.
+    """
+
```

Comment:
I suggest to add a generic `__call__` method without a specific signature (to avoid mypy errors in subclasses)
```suggestion

__call__: Callable
```

Reply:
As it is right now a `GTCallable` does not need to be callable directly, so I think we should skip this for now. The fact that the signature of `__call__` to a field operator does not match it's `__gt_type__` signature (we should fix this at some point) puts further insult to inquiry.

Reply:
It is impossible to define a generic field-operator or protocol whose `__call__` signature matches the `__gt_type__` signature of the object without dependent types, so we can already forget about having a fully correct solution here.
However, I think the protocol definition should state clearly that this object has to be callable (for embedded execution) even if right now.actual implementations raise a `NotImplementedError`

Reply:
I'm not completely sure I correctly understand what you mean here, but my point is that a field operator has a `__gt_type__` signature whose return type is a field, while it's `__call__` signature has return type void. Solving this is a question of using a different syntax inside programs. I've added a comment with respect to `__call__`.